### ISSUE TITLE
simple_lstm 训练准确率低

### ISSUE THREAD

**Issue by majiayuanyuan** [PaddlePaddle/Paddle#5458]
seq2seq 的10w分类预测，在集群上训练，用了100个节点，训练了24小时，准确率只有0.6%，用了三层的lstm，效果仍旧没有提升，请问是我的网络结构有问题吗？可以更换其他网络结构来提升效果吗？

我的输入数据类似如下格式：
word=[[1,2,3],[4,5],[5,6]] ;
bias = [[0.0,1.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0],
[0.0,1.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0],
[0.0,1.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0]]
lable=[[2,3,4],[5,6],[6,7]] (这里假设batch_size =3)，

具体网络结构如下:
```python
	data = paddle.layer.data("word",
	paddle.data_type.integer_value_sequence(input_dim))
	input_emb = paddle.layer.embedding(input=data, size=emb_dim)
	bias = paddle.layer.data("bias",
	paddle.data_type.dense_vector_sequence(9))
	concat = paddle.layer.concat(input=[bias, input_emb])
	for i in range(stacked_num):
	rnn_cell = paddle.networks.simple_lstm(
	input=bn if i else concat, size=hid_dim)
	bn = paddle.layer.batch_norm(input=rnn_cell, act=paddle.activation.Relu())

	output = paddle.layer.fc(
	input=[bn], size=class_dim, act=paddle.activation.Softmax())

	lbl = paddle.layer.data("label", paddle.data_type.integer_value_sequence(class_dim))
	cost = paddle.layer.classification_cost(input=output, label=lbl)
	parameters = paddle.parameters.create(cost)

	adam_optimizer = paddle.optimizer.Adam(
	learning_rate=0.005,
	learning_rate_schedule="pass_manual",
	learning_rate_args="0:1.0,1:0.5,2:0.2,3:0.1,4:0.05",
	regularization=paddle.optimizer.L2Regularization(rate=1e-3),
	model_average=paddle.optimizer.ModelAverage(average_window=0.5))

	trainer = paddle.trainer.SGD(
	cost=cost,
	extra_layers=paddle.evaluator.auc(input=prob, label=label),
	parameters=parameters,
	update_equation=adam_optimizer)

	feeding = {"word": 0, "label": 1, "bias": 2}
	trainer.train(
	reader=train_reader,
	event_handler=_event_handler,
	feeding=feeding,
	num_passes=num_pass)
 ```

**Comment by majiayuanyuan** [comment]
seq2seq 的10w分类预测，在集群上训练，用了100个节点， 训练了24小时，准确率只有0.6%， 试了3层的simple_lstm, 准确率依旧没有提升，请问是我的网络搭建有问题吗？ 或者有哪些效果更好的网络？
我的输入数据类似如下格式： 
word=[[1,2,3],[4,5],[5,6]];
bias = [[0.0,1.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0],
[0.0,1.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0],
[0.0,1.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0] ];
label=[[2,3,4],[5,6],[6,7]] (这里假设batch_size =3)
具体网络结构如下：
data = paddle.layer.data("word",
                         paddle.data_type.integer_value_sequence(input_dim))
input_emb = paddle.layer.embedding(input=data, size=emb_dim)

bias = paddle.layer.data("bias",
        paddle.data_type.dense_vector_sequence(9))
#reshape_bias = paddle.layer.seq_reshape(input=bias, reshape_size=9)
concat = paddle.layer.concat(input=[bias, input_emb])
for i in range(stacked_num):
    rnn_cell = paddle.networks.simple_lstm(
            input=bn if i else concat, size=hid_dim)
    bn = paddle.layer.batch_norm(input=rnn_cell, act=paddle.activation.Relu())

output = paddle.layer.fc(
        input=[bn], size=class_dim, act=paddle.activation.Softmax())

lbl = paddle.layer.data("label", paddle.data_type.integer_value_sequence(class_dim))
cost = paddle.layer.classification_cost(input=output, label=lbl)


 parameters = paddle.parameters.create(cost)
 # create optimizer
 adam_optimizer = paddle.optimizer.Adam(
     learning_rate=0.005,
     learning_rate_schedule="pass_manual",
     learning_rate_args="0:1.0,1:0.5,2:0.2,3:0.1,4:0.05",
     regularization=paddle.optimizer.L2Regularization(rate=1e-3),
     model_average=paddle.optimizer.ModelAverage(average_window=0.5))
 # create trainer
 trainer = paddle.trainer.SGD(
     cost=cost,
     extra_layers=paddle.evaluator.auc(input=prob, label=label),
     parameters=parameters,
     update_equation=adam_optimizer)
 # begin training network
 feeding = {"word": 0, "label": 1, "bias": 2}
 trainer.train(
     reader=train_reader,
     event_handler=_event_handler,
     feeding=feeding,
     num_passes=num_pass)

**Comment by majiayuanyuan** [comment]
seq2seq 的10w分类预测，在集群上训练，用了100个节点，训练了24小时，准确率只有0.6%，用了三层的lstm，效果仍旧没有提升，请问是我的网络结构有问题吗？可以更换其他网络结构来提升效果吗？

我的输入数据类似如下格式： 
word=[[1,2,3],[4,5],[5,6]] ; 
bias = [[0.0,1.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0],
[0.0,1.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0],
[0.0,1.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0]]
lable=[[2,3,4],[5,6],[6,7]] (这里假设batch_size =3)，
具体网络结构如下，
data = paddle.layer.data("word",
                         paddle.data_type.integer_value_sequence(input_dim))
input_emb = paddle.layer.embedding(input=data, size=emb_dim)

bias = paddle.layer.data("bias",
        paddle.data_type.dense_vector_sequence(9))
#reshape_bias = paddle.layer.seq_reshape(input=bias, reshape_size=9)
concat = paddle.layer.concat(input=[bias, input_emb])
for i in range(stacked_num):
    rnn_cell = paddle.networks.simple_lstm(
            input=bn if i else concat, size=hid_dim)
    bn = paddle.layer.batch_norm(input=rnn_cell, act=paddle.activation.Relu())

output = paddle.layer.fc(
        input=[bn], size=class_dim, act=paddle.activation.Softmax())

lbl = paddle.layer.data("label", paddle.data_type.integer_value_sequence(class_dim))
cost = paddle.layer.classification_cost(input=output, label=lbl)


parameters = paddle.parameters.create(cost)
# create optimizer
adam_optimizer = paddle.optimizer.Adam(
    learning_rate=0.005,
    learning_rate_schedule="pass_manual",
    learning_rate_args="0:1.0,1:0.5,2:0.2,3:0.1,4:0.05",
    regularization=paddle.optimizer.L2Regularization(rate=1e-3),
    model_average=paddle.optimizer.ModelAverage(average_window=0.5))
# create trainer
trainer = paddle.trainer.SGD(
    cost=cost,
    extra_layers=paddle.evaluator.auc(input=prob, label=label),
    parameters=parameters,
    update_equation=adam_optimizer)
# begin training network
feeding = {"word": 0, "label": 1, "bias": 2}
trainer.train(
    reader=train_reader,
    event_handler=_event_handler,
    feeding=feeding,
    num_passes=num_pass)

**Comment by majiayuanyuan** [comment]
seq2seq 的10w分类预测，在集群上训练，用了100个节点，训练了24小时，准确率只有0.6%，用了三层的lstm，效果仍旧没有提升，请问是我的网络结构有问题吗？可以更换其他网络结构来提升效果吗？

我的输入数据类似如下格式：
word=[[1,2,3],[4,5],[5,6]] ;
bias = [[0.0,1.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0],
[0.0,1.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0],
[0.0,1.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0]]
lable=[[2,3,4],[5,6],[6,7]] (这里假设batch_size =3)，

具体网络结构如下:

	data = paddle.layer.data("word",
	paddle.data_type.integer_value_sequence(input_dim))
	input_emb = paddle.layer.embedding(input=data, size=emb_dim)
	bias = paddle.layer.data("bias",
	paddle.data_type.dense_vector_sequence(9))
	concat = paddle.layer.concat(input=[bias, input_emb])
	for i in range(stacked_num):
	rnn_cell = paddle.networks.simple_lstm(
	input=bn if i else concat, size=hid_dim)
	bn = paddle.layer.batch_norm(input=rnn_cell, act=paddle.activation.Relu())

	output = paddle.layer.fc(
	input=[bn], size=class_dim, act=paddle.activation.Softmax())

	lbl = paddle.layer.data("label", paddle.data_type.integer_value_sequence(class_dim))
	cost = paddle.layer.classification_cost(input=output, label=lbl)
	parameters = paddle.parameters.create(cost)

	adam_optimizer = paddle.optimizer.Adam(
	learning_rate=0.005,
	learning_rate_schedule="pass_manual",
	learning_rate_args="0:1.0,1:0.5,2:0.2,3:0.1,4:0.05",
	regularization=paddle.optimizer.L2Regularization(rate=1e-3),
	model_average=paddle.optimizer.ModelAverage(average_window=0.5))

	trainer = paddle.trainer.SGD(
	cost=cost,
	extra_layers=paddle.evaluator.auc(input=prob, label=label),
	parameters=parameters,
	update_equation=adam_optimizer)

	feeding = {"word": 0, "label": 1, "bias": 2}
	trainer.train(
	reader=train_reader,
	event_handler=_event_handler,
	feeding=feeding,
	num_passes=num_pass)

**Comment by lcy-seso** [comment]
这个不是 seq2seq ，只是基础的分类问题。10w 分类问题类别数目不算少了，大致可以类比于小规模的语言模型这样的任务，不知道具体任务是什么。10w 类的先验分布是什么样的。

准确率 0.6 对 10w类分类，虽然不算高，但也不是随机猜测。有没有更多的任务背景，训练情况（收敛曲线，收敛速率之类）的信息。

**Comment by lcy-seso** [comment]
有一个地方我没太懂，从上面的例子看， embedding 出来是三个序列，bias 是一个有三个时间步的序列？concat word embedding 和 bias 不会报错吗？

上面的配置里面，LSTM 的input-to-hidden加BN，激活用RELU，对我个人来看，都需要根据具体任务来判断是否最后的配置方法，比较令人怀疑。

**Comment by majiayuanyuan** [comment]
对 ，这里面实际就是分类问题，这里 word 是用户的一个点击序列，bias 是用户的 的基本信息（年龄，性别 one-hot 成9纬），每个序列 中的每个 元素 进行embeding 最后和用户的bias信息进行拼接 ，进行预测  ，之所以用这个结构，是因为之前用tf 的lstm 进行预测，tf 的lstm 有一个状态的初始设置，用的就是bias 信息 ，paddle  这里没有，所以尝试的这种方法，用 tf 预测时，准确率可以达到2.2%， 所以觉得0.6% 这个准确率有些低

**Comment by majiayuanyuan** [comment]
不使用用户的bias信息 进行拼接，我也尝试过，效果和这个差不多
